frontend/src/plugins/impl/FileBrowserPlugin.tsx

```diff
@@ -0,0 +1,279 @@
+/* Copyright 2024 Marimo. All rights reserved. */
+import { z } from "zod";
+import { Table, TableCell, TableRow } from "@/components/ui/table";
+import { createPlugin } from "../core/builder";
+import { useState } from "react";
+import { useAsyncData } from "@/hooks/useAsyncData";
+import { rpc } from "../core/rpc";
+import { Checkbox } from "@/components/ui/checkbox";
+import { Folder } from "lucide-react";
+import { toast } from "@/components/ui/use-toast";
+import { NativeSelect } from "@/components/ui/native-select";
+
+/**
+ * Arguments for a file browser component.
+ *
+ * @param initialPath - the path to display on component render
+ * @param filetypes - filter directory lists by file types
+ * @param multiple - whether to allow the user to select multiple files
+ * @param label - label for the file browser
+ * @param restrictNavigation - whether to prevent the user from accessing
+ * directories outside the set path
+ */
+interface Data {
+  initialPath: string;
+  filetypes: string[];
+  multiple: boolean;
+  label: string | null;
+  restrictNavigation: boolean;
+}
+
+/**
+ * @param id - File id
+ * @param path - File path
+ * @param name - File name
+ * @param is_directory - Whether file is a directory or not
+ * @param is_marimo_file - Whether file is a marimo file or not
+ */
+interface FileInfo {
+  id: string;
+  path: string;
+  name: string;
+  is_directory: boolean;
+  is_marimo_file: boolean;
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type PluginFunctions = {
+  list_directory: (req: { path: string; filetypes: string[] }) => Promise<{
+    files: FileInfo[];
+  }>;
+};
+
+type S = FileInfo[];
+
+export const FileBrowserPlugin = createPlugin<S>("marimo-file-browser")
+  .withData(
+    z.object({
+      initialPath: z.string(),
+      filetypes: z.array(z.string()),
+      multiple: z.boolean(),
+      label: z.string().nullable(),
+      restrictNavigation: z.boolean(),
+    }),
+  )
+  .withFunctions<PluginFunctions>({
+    list_directory: rpc
+      .input(
+        z.object({
+          path: z.string(),
+          filetypes: z.array(z.string()),
```

Comment:
we should already know this on the backend (and dont need to pass it)

Reply:
Ah will do. I think I was going to add a way for the user to set the filetype filter from the UI, then decided not to but forgot to update the args

Reply:
oh thats a good idea - we can add that later too